merkletree.go

```diff
@@ -0,0 +1,274 @@
+		},
+	}
+
+	rightBranch := &emptyNode{
+		node: node{
+			level: 1,
+		},
+	}
+	root := &interiorNode{
+		node: node{
+			parent: nil,
+			level:  0,
+		},
+		leftChild:  leftBranch,
+		rightChild: rightBranch,
+		leftHash:   nil,
+		rightHash:  nil,
+	}
+	leftBranch.parent = root
+	rightBranch.parent = root
+
+	m := &MerkleTree{
+		treeNonce: treeNonce,
+		salt:      salt,
+		privKey:   privKey,
+		pubKey:    pubKey,
+		policies:  policies,
+		root:      root,
+	}
+	return m
+}
+
+func (h *History) LookUp(key string) (MerkleNode, []LookUpProofNode, error) {
+	str := h.currentSTR
+	return lookUp(key, str)
+}
+
+func (h *History) LookUpInEpoch(key string, ep int64) (MerkleNode, []LookUpProofNode, error) {
+	str := h.GetSTR(ep)
+	return lookUp(key, str)
+}
+
+func lookUp(key string, str *SignedTreeRoot) (MerkleNode, []LookUpProofNode, error) {
+	lookupIndex := computePrivateIndex(key)
+	position := 0
+	var nodePointer interface{}
+	nodePointer = str.treeRoot
+	var proof []LookUpProofNode
+
+	for {
+		if _, ok := nodePointer.(*userLeafNode); ok {
+			// reached to a leaf node
+			break
+		}
+		if _, ok := nodePointer.(*emptyNode); ok {
+			// reached to an empty branch
+			break
+		}
+		proof = append(proof, nodePointer.(*interiorNode))
+		direction := util.GetNthBit(lookupIndex, position)
+		if direction {
+			nodePointer = nodePointer.(*interiorNode).rightChild
+		} else {
+			nodePointer = nodePointer.(*interiorNode).leftChild
+		}
+		position++
+	}
+
+	if nodePointer == nil {
+		return nil, nil, ErrInvalidTree
+	}
+	switch nodePointer.(type) {
+	case *userLeafNode:
+		proof = append(proof, nodePointer.(*userLeafNode))
+		if nodePointer.(*userLeafNode).key == key {
+			return nodePointer.(*userLeafNode), proof, nil
+		}
+		// reached a different leaf with a matching prefix
+		// return nil and a auth path including the leaf node
+		return nil, proof, nil
+	case *emptyNode:
+		proof = append(proof, nodePointer.(*emptyNode))
+		return nil, proof, nil
+	}
+	return nil, nil, ErrInvalidTree
+}
+
+func (m *MerkleTree) Set(key string, value []byte) error {
+	index := computePrivateIndex(key)
+	toAdd := userLeafNode{
+		key:        key,
+		value:      value,
+		index:      index,
+		commitment: crypto.Digest(m.salt, []byte(key), value),
+	}
+
+	return m.insertNode(index, &toAdd)
+}
+
+// Private Index calculation function
+// would be replaced with Ismail's VRF implementation
+func computePrivateIndex(key string) []byte {
+	return crypto.Digest([]byte(key))
+}
+
+func (m *MerkleTree) insertNode(key []byte, toAdd *userLeafNode) error {
+	position := 0
+	var nodePointer interface{}
+	nodePointer = m.root
+
+insertLoop:
+	for {
+		toAdd.level++
+		switch nodePointer.(type) {
+		case *userLeafNode:
+			// reached a "bottom" of the tree.
+			// add a new interior node and push the previous leaf down
+			// then continue insertion
+			currentNodeUL := nodePointer.(*userLeafNode)
+			if currentNodeUL.parent == nil {
+				return ErrInvalidTree
+			}
```

Comment:
> Any error means a problem with our logic in the implementation.

I'm not sure this will be true in 100% of the cases. But I see your point about errors vs bugs.